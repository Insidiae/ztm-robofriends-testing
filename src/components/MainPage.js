import React from "react";

import Header from "./Header";
import SearchBox from "./SearchBox";
import Scroll from "./Scroll";
import CardList from "./CardList";

function MainPage({ onSearchChange, searchfield, robots, error, isPending }) {
  const filteredRobots = robots.length
    ? robots.filter((robot) => {
        return robot.name.toLowerCase().includes(searchfield.toLowerCase());
      })
    : [];

  return (
    <div className="tc">
      <Header />
      {/* <button onClick={() => setCount(count + 1)}>Click Me!</button> */}
      <SearchBox searchChange={onSearchChange} />
      <Scroll>
        {error ? (
          <>
            <h1>Something went wrong.</h1>
            <p>Error: {error.message}</p>
          </>
        ) : isPending ? (
          <h1>Loading</h1>
        ) : (
          <CardList robots={filteredRobots} />
        )}
      </Scroll>
    </div>
  );
}

export default MainPage;
