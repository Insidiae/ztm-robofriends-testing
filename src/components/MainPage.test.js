import React from "react";
import { render, screen } from "@testing-library/react";

import MainPage from "./MainPage";

describe("<MainPage />", () => {
  it("renders correctly", () => {
    const mockProps = {
      onSearchChange: jest.fn(),
      searchField: "",
      robots: [],
      error: "",
      isPending: false,
    };
    const { container } = render(<MainPage {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  it("displays loading text", () => {
    const mockProps = {
      onSearchChange: jest.fn(),
      searchField: "a",
      robots: [],
      error: "",
      isPending: true,
    };

    render(<MainPage {...mockProps} />);

    expect(screen.getByText("Loading")).toBeInTheDocument();
  });

  it("filters robots", () => {
    const mockProps = {
      onSearchChange: jest.fn(),
      searchField: "a",
      robots: [],
      error: "",
      isPending: false,
    };

    render(<MainPage {...mockProps} />);

    expect(screen.queryByAltText("robots")).toBeNull();
  });

  it("filters robots correctly", () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [
        {
          id: 1,
          name: "Leanne Graham",
          username: "Bret",
          email: "Sincere@april.biz",
        },
      ],
      searchfield: "Leanne",
      error: "",
      isPending: false,
    };

    render(<MainPage {...mockProps} />);

    expect(screen.getByText("Leanne Graham")).toBeInTheDocument();
  });

  it("filters robots correctly 2", () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [
        {
          id: 1,
          name: "Leanne Graham",
          username: "Bret",
          email: "Sincere@april.biz",
        },
      ],
      searchfield: "Xavier",
      error: "",
      isPending: false,
    };

    render(<MainPage {...mockProps} />);

    expect(screen.queryByText("Leanne Graham")).toBeNull();
  });

  it("displays error correctly", () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [],
      searchfield: "nope",
      error: "Oh no, an error!",
      isPending: false,
    };

    render(<MainPage {...mockProps} />);

    expect(screen.getByText("Something went wrong.")).toBeInTheDocument();
  });
});
