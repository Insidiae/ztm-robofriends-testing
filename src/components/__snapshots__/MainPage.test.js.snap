// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainPage /> renders correctly 1`] = `
<div>
  <div
    class="tc"
  >
    <div>
      <h1
        class="f1"
      >
        RoboFriends
      </h1>
      <button
        color="red"
        id="counter"
      >
        Count: 
        0
      </button>
    </div>
    <div
      class="pa2"
    >
      <input
        aria-label="Search Robots"
        class="pa3 ba b--green bg-lightest-blue"
        placeholder="search robots"
        type="search"
      />
    </div>
    <div
      style="overflow: scroll; border: 5px solid black; height: 800px;"
    >
      <div />
    </div>
  </div>
</div>
`;
